piff/writers.py

```diff
@@ -0,0 +1,257 @@
+
+    A `Writer` subclass is always paired with a `Reader` subclass, and the
+    methods of `Reader` each directly correspond to a method of `Writer`.
+
+    All `Reader` and `Writer` methods take a ``name`` argument that is
+    associated with the low-level data structured being saved.  When writing,
+    an object can chose not to write a subobject at all with a given name, and
+    then check to see if the corresponding `Reader` method returns `None`, but
+    a `Reader` cannot be used to see what type of data structure was saved with
+    a given name; if this can change, the write implementation must store this
+    explicitly and use it when invoking the `Reader`.
+    """
+
+    @staticmethod
+    @contextmanager
+    def open(file_name):
+        """Return a context manager that yields a `Writer` appropriate for the
+        given filename.
+
+        :param filename:   Name of the file to open (`str`).
+
+        :returns:  A context manager that yields a `Writer`.
+        """
+        _, ext = os.path.splitext(file_name)
+        if ext == ".fits" or ext == ".piff":
+            with FitsWriter._open(file_name) as writer:
+                yield writer
+            return
+        else:
+            raise NotImplementedError(f"No writer for extension {ext!r}.")
+
+    def write_struct(self, name, struct):
+        """Write a simple flat dictionary.
+
+        :param name:        Name used to save this struct.
+        :param struct:      A `dict` with `str` keys and `int`, `float`, `str`,
+                            `bool`, or `None` values.
+        """
+        raise NotImplementedError()
+
+    def write_table(self, name, table, metadata=None):
+        """Write a table via a numpy array.
+
+        :param name:      Name used to save this table.
+        :param metadata:  A `dict` of simple metadata to save with the table.
+                          Keys must be `str` (case may not be preserved) and
+                          values must be `int`, `float`, `str`, or `bool`.
+        """
+        raise NotImplementedError()
+
+    def write_array(self, name, array, metadata=None):
+        """Write a numpy array that does not have a structured dtype.
+
+        :param name:      Name used to save this array.
+        :param metadata:  A `dict` of simple metadata to save with the array.
+                          Keys must be `str` (case may not be preserved) and
+                          values must be `int`, `float`, `str`, or `bool`.
+        """
+        raise NotImplementedError()
+
+    def write_wcs_map(self, name, wcs_map, pointing):
+        """Write a regular a map of WCS objects and an optoinal pointing coord.
+
+        :param name:      Name used to save this struct in `Writer.write_array`.
+        :param wcs_map:   A `dict` mapping `int` chipnum to `galsim.BaseWCS`.
+        :param pointing:  A `galsim.CelestialCoord`, or `None`.
+        :param logger:    A logger object for logging debug info.
+        """
```

Comment:
I've moved the WCS reading and writing in the `PSF` class into the implementation of this method and its `Reader` counterpart, instead of leaving them in `PSF` and having them call `write_table` and `read_table`.  That's because I envision reimplementing this specifically for Rubin since we'll be saving the WCSs ourselves elsewhere in the same file as the PSF most of the time, and we might want to just have a string/int pointer to that written by the writer and dereferenced by the reader.